spec/datadog/integration_spec.rb

```diff
@@ -22,23 +22,31 @@ def wait_for_tracer_sent
     end
 
     context 'for threads' do
-      let(:original_threads) { Thread.list }
-      let!(:original_threads_inspect) { inspect_threads(original_threads) } # Store result as stack trace will change
-      let(:threads) { Thread.list }
+      let!(:original_threads) { Thread.list }
+      let(:start_tracer) do
+        Datadog::Tracing.trace('test.op') {}
+
+        new_threads = Thread.list - original_threads
+        tracer_threads.concat(new_threads)
+      end
+      let(:tracer_threads) { [] }
 
       def inspect_threads(threads)
-        threads.map.with_index { |t, idx| "#{idx}=#{t.backtrace}" }.join(';')
+        threads.map.with_index { |t, idx| "#{idx}=#{t.object_id}:#{t.backtrace}" }.join(';')
       end
 
+      subject(:shutdown) { Datadog.shutdown! }
+
       it 'closes tracer threads' do
         start_tracer
         wait_for_tracer_sent
 
         shutdown
 
-        expect(threads.size).to eq(original_threads.size),
-          "Expected #{original_threads.size} threads (#{original_threads_inspect}), "\
-                  "got #{threads.size} threads (#{inspect_threads(threads)})"
+        post_shutdown_threads = Thread.list
+
+        expect(post_shutdown_threads & tracer_threads).to be_empty,
+          "Tracer threads not terminated: #{inspect_threads(post_shutdown_threads)}"
```

Comment:
So just to make sure I'm following the change here: The previous flaky test was checking that between the start of the test and after calling `shutdown`, there were no threads remaining.

The new version is instead checking that between the start of the test, and after calling `Datadog::Tracing.trace` no new threads remain. Thus, the `wait_for_tracer_sent` and `shutdown` parts aren't covered anymore. And presumably the flakiness was coming from there and since we're not counting them, the test will pass?

Reply:
As I understand it, the previous iteration of this test was simply checking the number of threads present in `Threads.list` before and after the tracer started and shutdown. Also, as you pointed out in my refactor, the old test was lazily evaluating the instantiation of `original_threads`.

I refactored the test to check the specific threads created when the tracer starts are no longer present after shutdown. My thesis is that unrelated threads may have been in `Threads.list` at the time of observation when the test executes, causing the flakiness. The `wait_for_tracer_sent` and `shutdown` are still executed and necessary for the test to pass in the refactored version.

Reply:
> I refactored the test to check the specific threads created when the tracer starts are no longer present after shutdown.

Indeed! What I'm pointing out is -- I'm not sure _when_ the extra threads that caused the test to be flaky were created.

So if those threads happened in the `wait_for_tracer_sent` or `shutdown` phases, we now ignore them, which is presumably why the flakiness went away -- right?

> as you pointed out in my refactor, the old test was lazily evaluating the instantiation of original_threads.

Nitpick xD but this was actually not the case! The old test had a `let!(:original_threads_inspect)` which called the `original_threads`, making it so that indirectly, `original_threads` was a `let!`. The issue I pointed out actually only happened after the inspect was removed.

Reply:
I dug into what the extra threads were and it turns out it's Ruby's [Timeout](https://ruby-doc.org/stdlib-3.1.0/libdoc/timeout/rdoc/Timeout.html) module. 

Adding `Timeout.ensure_timeout_thread_created if Timeout.respond_to?(:ensure_timeout_thread_created)` to our spec_helper ensures the thread is there before the test is run so it is included in `original_threads` and therefore should stop the flakiness.

I mostly reverted my test changes as @marcotc pointed out it was testing only the threads created at startup are no longer present. But it's possible threads created after startup could still exist and my test would incorrectly pass.

Reply:
Nice!